src/chabochethermal.jl

```diff
@@ -0,0 +1,1050 @@
+
+Compute `n = ∂f/∂σ`.
+
+`state` should contain `stress`, `R`, `X1`, `X2`, `X3`.
+`drivers` should contain `temperature`.
+`parameters` should contain `R0`, a function of temperature.
+
+Other properties of the structures are not used by this function.
+
+The return value is the symmetric rank-2 tensor `n`.
+"""
+function yield_jacobian(state::GenericChabocheThermalVariableState{<:Real},
+                        drivers::GenericChabocheThermalDriverState{<:Real},
+                        parameters::GenericChabocheThermalParameterState{<:Real})
+    # We only need ∂f/∂σ, so let's compute only that to make this run faster.
+    #
+    # # TODO: The `gradient` wrapper of `Tensors.jl` is nice, but it doesn't tag its Dual.
+    # #
+    # # When using `Tensors.gradient` in `yield_jacobian` (n = ∂f/∂σ), the
+    # # differentiation of `yield_criterion` with respect to stress doesn't work
+    # # when computing the temperature jacobian for the residual function, which
+    # # needs ∂n/∂θ = ∂²f/∂σ∂θ. `ForwardDiff` fails to find an ordering for the
+    # # `Dual` terms (the temperature Dual having a tag, but the stress Dual not).
+    # #
+    # # Using `ForwardDiff.jacobian` directly, both Duals are tagged, so this works.
+    #
+    # @unpack stress, R, X1, X2, X3 = state
+    # function f(stress::Symm2{<:Real})
+    #     state = GenericChabocheThermalVariableState{eltype(stress)}(stress=stress,
+    #                                                                 R=R,
+    #                                                                 X1=X1,
+    #                                                                 X2=X2,
+    #                                                                 X3=X3)
+    #     return yield_criterion(state, drivers, parameters)
+    # end
+    # return gradient(f, stress)
+    @unpack stress, R, X1, X2, X3 = state
+    marshal(tensor::Symm2) = tovoigt(tensor)
+    unmarshal(x::AbstractVector{T}) where T <: Real = fromvoigt(Symm2{T}, x)
+    function f(x::AbstractVector{<:Real})  # x = stress
+        state = GenericChabocheThermalVariableState{eltype(x)}(stress=unmarshal(x),
+                                                               X1=X1,
+                                                               X2=X2,
+                                                               X3=X3,
+                                                               R=R)
+        return [yield_criterion(state, drivers, parameters)]::Vector
+    end
+    J = ForwardDiff.jacobian(f, marshal(stress))
+    # The result is a row vector, so drop the singleton dimension.
+    return unmarshal(J[1,:])
+end
+
+"""
+    overstress_function(state::GenericChabocheThermalVariableState{<:Real},
+                           drivers::GenericChabocheThermalDriverState{<:Real},
+                           parameters::GenericChabocheThermalParameterState{<:Real})
+
+Norton-Bailey type power law.
+
+`parameters` should contain `tvp`, `Kn` and `nn`.
```

Comment:
The current parameter containers are tied to the material type, but in the future, if we are able to modularize the 'behavior laws', should the parameters be contained within the initialized instances of the 'behavior laws' and the instances of initialized 'behaviors' contained in the material-container instead?

Reply:
Yes, definitely. Actually, that's precisely what I'm working on at the moment.

We'll need the modular structure to be able to test against Bari's results, while keeping the code maintainable. The aim is to provide both overstress viscoplastic as well as pure plastic variants of the model, which allows us to include those tests in the test suite.

There are some other issues with extending the original design that need to be thought out:

- Now that the material parameters are functions, `dparameters` no longer makes sense. `Materials.jl` doesn't use it. Because, as we discussed offline, other parts of JuliaFEM don't need it either, I suggest we drop `dparameters` and bump the major version of `Materials.jl`.
- The multi-headed hydra that is `drivers`/`ddrivers` and `variables`/`variables_new`.
  - In the context of the original design, it's fine. I see the convenience - the integrator function needs only one argument, namely the material struct. But as we generalize, we need to think about how to keep the codebase easily understandable and maintainable.
  - Conceptually, `drivers` (*t*, *ε*, and the temperature *θ*) can be thought of as part of the material state.
  - However, `ddrivers` is a delta of that state. Moreover, it's tentative in the sense that one needs to apply `integrate_material!` (which may fail, e.g. if *Δt* is too large), and then commit the changes with `update_material!`, before that change will become part of the current material state. For now, however, I'll keep `ddrivers` in the top-level material model. This can be changed later if desired.
  - `variables_new` is essentially a duplicate of `variables`, except for the obvious difference that it holds the new (yet uncommitted) data. Having a rollback capability - which is what having `variables_new` provides - is in itself essential, so this is good. The design question is whether it's better to have one multi-headed hydra (like now), or two separate material instances. For now, I'm postponing the quest for the perfect design in this regard. I'm keeping `variables_new`, but moving it to the submodels. Each submodel that has a `variables` field will also have `variables_new`. There will also be a walker that can marshal/unmarshal variables from/to these nested model hierarchies. As discussed offline, we can generate the mapping between fields and marshaled vector indices at instantiation time, so the runtime cost will be minimal.

But without showing the actual new design, my points above are rather academic (in the programmer sense) - I'll post my draft soon to facilitate discussion. I think 100-200 lines of struct definitions is a rather compact and flexible format to present the ideas. ;)

Reply:
I fully agree.

To comment on `dparameters` - it was never used before but was designed as a placeholder for situations like this, that we have e.g. temperature-dependent Young's modulus. It was not decided which part of the code should take care of the interpolation or updation of the parameter-values. In JuliaFEM the `Fields`-system actually currently has the parameter definitions (both spatial, temporal, and other field-dependencies such as temperature) stored in the integration points/nodes (can't remember which). So in this case the parameter values and changes to them would come from the FEM-solver. But then again we would need the jacobians against all drivers. I hope this helps to see the bigger picture.

This was a design choice by @ahojukka5 and it was done to be as flexible as possible, but we have to remember that it also dates to the times where materials considered were mainly elastic (and material response/jacobian was defined on the solver level explicitly). We have later realized that the current `Fields` system might not be the best design at least in terms of memory usage and I believe there have been some ideas cooking on changing that.

Reply:
Thanks for explaining the rationale behind the original design. Yes, it was fine against that background - the reason we need to change it now is precisely that we are now in the process of deciding where to handle the temperature dependence w.r.t. `Materials.jl` itself.

Having the temperature dependence in the material model allows us to provide jacobians if needed, as well as allows for standalone single material point simulations in scenarios where the temperature dependence is important.

I agree that `Materials.jl` is useful by itself for its single material point solvers, and indeed it is generic enough that (given a suitable adaptor to translate between the APIs) it can work with any FEM.

I'd be happy to hear those ideas! Could be useful in deciding how to go forward here.